### PR TITLE
Allow creating wxUILocale for any locale and add wxUILocale::CompareStrings()

### PR DESCRIPTION
**Pull request by vadz**
This makes it possible to perform locale-dependent operations portably, without affecting the current locale.

It's mostly useful for the applications working with texts in some fixed languages that can be different from the current UI locale.

### CODE REVIEW

**Comment by PBfordev** [comment]
I am sorry for the stupid question but still: What is comparing strings supposedly containing "non-English" characters good for?

I would assume that string comparisons would be mostly used for sorting a list of strings to be presented to the user in the alphabetical order as understood by a common user; however, in such cases collate instead compare should be used for this to actually work?

It may be just a bias due to my language (Czech), where using compare results into unpleasantly surprising behaviour. For example, characters with diacritics are not ordered as expected (e.g., 'č' between 'c' and 'd') but go after 'z' which is unacceptable. Is Czech really is exceptional in this regard (see also [here](https://groups.google.com/g/wx-users/c/DrA7vmNPYJE/m/Xm3MYK7EAgAJ))?

**Comment by vadz** [comment]
> I am sorry for the stupid question but still: What is comparing strings supposedly containing "non-English" characters good for?

For sorting strings in order according to this language rules.
 
> I would assume that string comparisons would be mostly used for sorting a list of strings to be presented to the user in the alphabetical order as understood by a common user; however, in such cases collate instead compare should be used for this to actually work?

Maybe it's my question which is stupid, but what difference do you see between comparing and collating? IMO "compare" is a more clear term, but for me is supposed to do the same thing, e.g. it's literally implemented using `wcscoll()` function on POSIX systems.

> It may be just a bias due to my language (Czech), where using compare results into unpleasantly surprising behaviour. For example, characters with diacritics are not ordered as expected (e.g., 'č' between 'c' and 'd') but go after 'z' which is unacceptable. Is Czech really unique in this regard (see also [here](https://groups.google.com/g/wx-users/c/DrA7vmNPYJE/m/Xm3MYK7EAgAJ)) and compare can actually be used for sorting in other languages?
> 
> **EDIT**
> Based on the [information in wikipedia](https://en.wikipedia.org/wiki/German_orthography#Sorting), I think compare for sorting strings cannot be used at least for German either, at least if we are not considering seemingly pretty rare case #3.2 there.

As the unit tests added in this PR show, German sorting does work according to DIN 5007-1, as expected. I didn't test Czech yet, but I'd expect sorting for it to work properly too, i.e. with `"c" < "č" < "d"`. Do you mean that you've tested it and that it doesn't work for you with the existing code?

**Comment by vadz** [comment]
FWIW I've just tested this diff:
```diff
diff --git a/tests/intl/intltest.cpp b/tests/intl/intltest.cpp
index a81f6e5232..5fa2ee2fbe 100644
--- a/tests/intl/intltest.cpp
+++ b/tests/intl/intltest.cpp
@@ -351,6 +351,14 @@ static inline bool CheckSupported(const wxUILocale& loc, const char* desc)
         CHECK( l.CompareStrings(u8("ä"), "ae") == 1 );
         CHECK( l.CompareStrings(u8("ö"), "z" ) == 1 );
     }
+
+    SECTION("Czech")
+    {
+        const wxUILocale l("cs");
+        REQUIRE( l.IsSupported() );
+        CHECK( l.CompareStrings("c", u8("č")) == -1 );
+        CHECK( l.CompareStrings(u8("č"), "d") == -1 );
+    }
 }
 
 #endif // wxUSE_INTL
```
and it works as expected, i.e. all the tests pass.

**Comment by PBfordev** [comment]
I apologize, you are right and I am stupid, the code works.  I did my tests with this PR but it did not work for me. 

Unfortunately, I discarded my test code so the only explanation seems to be that I must have swapped the order of variables passed to `CompareStrings()`.

**Comment by vadz** [comment]
Right, and I'm very smart because I must have exchanged the order in them an even amount of times (and definitely much greater than 2) :-)

It's true that it's easy to make mistakes with this API, but it's consistent with everything else.

BTW, another possibility is that creating the locale failed: I initially tried "cz" (instead of "cs") and it didn't work neither because it silently falls back to "C" locale which, I think, is more useful than the alternative (which would be to do what, exactly?), but also error-prone.

**Comment by PBfordev** [comment]
> BTW, another possibility is that creating the locale failed

I am stupid but not that stupid to not test whether the locale was set successfully. ;) My code looked like this (except it must have had a bug):
```cpp
const wxUILocale uiLocale(wxLocaleIdent("cs").Region("CZ"));
wxString s1(L"\u010D"); // č
wxString s2(L"d");

wxLogDebug("uiLocale name: %s", uiLocale.GetName());
wxLogDebug("uiLocale.CompareStrings(%s, %s) = %d", s1, s2, uiLocale.CompareStrings(s1, s2, wxCompare_CaseInsensitive));
```

**Comment by vadz** [comment]
This is finally ready for review, I think. @csomor (please see the special test for Swiss German dedicated to you :-) and @vslavik, please have a look if you can -- but, of course, any comments from everybody else would be welcome too. TIA!

File: include/wx/localedefs.h [hunk 1]
```diff
@@ -108,6 +108,83 @@ struct WXDLLIMPEXP_BASE wxLanguageInfo
...
+// ----------------------------------------------------------------------------
+
+class WXDLLIMPEXP_BASE wxLocaleIdent
+{
+public:
+    // Leave language empty
+    wxLocaleIdent() { }
+
+    // Construct from language, i.e. a two-letter ISO 639-1 code (or a
+    // three-lette ISO 639-2 code if there is no ISO 639-1 code for this
```
Comment (PBfordev) [inline include/wx/localedefs.h:122]:
```suggestion
    // three-letter ISO 639-2 code if there is no ISO 639-1 code for this
```

Comment (PBfordev) [inline include/wx/localedefs.h:122]:
```suggestion
    // three-letter ISO 639-2 code if there is no ISO 639-1 code for this
```

File: include/wx/uilocale.h [hunk 2]
```diff
@@ -38,19 +45,31 @@ class WXDLLIMPEXP_BASE wxUILocale
...
+    bool IsSupported() const;
+
     // Get the platform-dependent name of the current locale.
     wxString GetName() const;
 
     // Query the locale for the specified information.
     wxString GetInfo(wxLocaleInfo index,
                      wxLocaleCategory cat = wxLOCALE_CAT_DEFAULT) const;
 
+    // Compares two strings, for a locale specified by wxLocaleIdent.
```
Comment (PBfordev) [inline include/wx/uilocale.h:62]: Does the "for a locale specified by wxLocaleIdent" part really belong there?

File: interface/wx/uilocale.h [hunk 3]
```diff
@@ -80,6 +99,49 @@ class wxUILocale
      */
     static const wxUILocale& GetCurrent();
 
+    /**
+        Create the locale corresponding to the given locale identifier.
```
Comment (PBfordev) [inline interface/wx/uilocale.h:103]: Supernitpick: We had a discussion about verb forms some time ago and it was agreed to use Creates/Sets/Returns... instead of Create/Set/Return..., see the first bullet [here](https://github.com/wxWidgets/wxWidgets/blob/master/docs/contributing/how-to-add-class-documentation.md#important-dos-and-donts).

File: interface/wx/uilocale.h [hunk 4]
```diff
@@ -119,3 +192,135 @@ class wxUILocale
...
+            See Language() for more detailed info.
+    */
+    wxLocaleIdent(const char* language);
+
+    /**
+        Set language.
+
+        Return reference to @this for method chaining.
+
+        See http://www.loc.gov/standards/iso639-2/php/English_list.php for the
```
Comment (PBfordev) [inline interface/wx/uilocale.h:238]:
```suggestion
        See https://www.loc.gov/standards/iso639-2/php/English_list.php for the
```

File: src/msw/uilocale.cpp [hunk 5]
```diff
@@ -65,18 +74,49 @@ void wxUseLCID(LCID lcid)
...
 
-// TODO-XP: Replace this with an implementation using GetLocaleInfoEx() when we
-// don't support XP any longer.
+wxString wxLocaleIdent::GetName() const
+{
+    // Construct name in right format:
+    // Windows: <language>-<script>-<REGION>
+
+    wxString name;
+    if (!m_language.empty())
```
Comment (PBfordev) [inline src/msw/uilocale.cpp:86]:
```suggestion
    if ( !m_language.empty() )
```

File: src/msw/uilocale.cpp [hunk 6]
```diff
@@ -65,18 +74,49 @@ void wxUseLCID(LCID lcid)
...
+    {
+        name << m_language;
+
+        if (!m_script.empty())
```
Comment (PBfordev) [inline src/msw/uilocale.cpp:90]:
```suggestion
        if ( !m_script.empty() )
```

File: src/msw/uilocale.cpp [hunk 7]
```diff
@@ -65,18 +74,49 @@ void wxUseLCID(LCID lcid)
...
+        {
+            name << "-" << m_script;
+        }
+
+        if (!m_region.empty())
```
Comment (PBfordev) [inline src/msw/uilocale.cpp:95]:
```suggestion
        if ( !m_region.empty() )
```

**Review by PBfordev — COMMENTED** [review]
Just few code-unrelated nitpicks.

File: src/osx/core/uilocale.mm [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+#if wxUSE_INTL
+
+#include "wx/uilocale.h"
+#include "wx/private/uilocale.h"
+
+#include "wx/osx/core/cfref.h"
+#include "wx/osx/core/cfstring.h"
+
+#include <Foundation/NSString.h>
+#include <Foundation/NSLocale.h>
```
Comment (csomor) [inline src/osx/core/uilocale.mm:30]: I'd always include objective-C headers with `#import`, only c/c++ headers with `#include`

**Review by csomor — CHANGES_REQUESTED** [review]

**Comment by csomor** [comment]
@vadz I've pushed some changes to the xcodeproject, the .cpp -> .mm rename also caught the common cpp version

File: src/osx/core/uilocale.mm [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+{
+
+// ----------------------------------------------------------------------------
+// wxUILocale implementation using Core Foundation
+// ----------------------------------------------------------------------------
+
+class wxUILocaleImplCF : public wxUILocaleImpl
+{
+public:
+    // Takes ownership of NSLocale.
```
[the remark this answers is not in the document]
Comment (vadz) [reply src/osx/core/uilocale.mm:73]: It's a private/internal class so I thought it would be ok to take ownership here, even if it's going to make the code slightly more complicated (because we'll need an extra `release` in `Create()`).

I was really surprised, but apparently there is no smart pointer for Objective C++ code in wx similar to `wxCFRef` -- or did I miss it?

File: include/wx/localedefs.h (hunk 1 above)
Comment (vadz) [reply include/wx/localedefs.h:122]: Thanks for noticing this! I wonder why this wasn't caught by our spelling check...

File: include/wx/uilocale.h (hunk 2 above)
Comment (vadz) [reply include/wx/uilocale.h:62]: No, it's a leftover from the original version in which it was static, thanks for noticing!

File: interface/wx/uilocale.h (hunk 3 above)
Comment (vadz) [reply interface/wx/uilocale.h:103]: I completely forgot about it, thanks.

File: interface/wx/uilocale.h (hunk 4 above)
Comment (vadz) [reply interface/wx/uilocale.h:238]: Good point, thanks.

File: src/msw/uilocale.cpp (hunk 6 above)
Comment (vadz) [reply src/msw/uilocale.cpp:90]: I had originally decided it wasn't worth fixing the style in the original submission, but let's do it if I'm not the only bothered by it.

File: src/osx/core/uilocale.mm (hunk 8 above)
Comment (vadz) [reply src/osx/core/uilocale.mm:30]: Fixed, thanks.

File: src/osx/core/uilocale.mm (hunk 9 above)
[the remark this answers is not in the document]
Comment (csomor) [reply src/osx/core/uilocale.mm:73]: > It's a private/internal class so I thought it would be ok to take ownership here, even if it's going to make the code slightly more complicated (because we'll need an extra `release` in `Create()`).

the call you use currently there `[NSLocale localeWithLocaleIdentifier:...]` is a class factory method, you are just supposed to use the object instance with these kind of calls within the method, not release if after https://developer.apple.com/library/archive/documentation/General/Conceptual/CocoaEncyclopedia/ClassFactoryMethods/ClassFactoryMethods.html  . So if you intend to keep it longer, you have to retain it.

> I was really surprised, but apparently there is no smart pointer for Objective C++ code in wx similar to `wxCFRef` -- or did I miss it?

there is `template <class T> class wxNSObjRef`, since CFLocale and NSLocale are toll-free bridged you could also use a wxCFRef. When I have a obj-c++ class, then I'm using the native properties for these attributes.

[the remark this answers is not in the document]
Comment (csomor) [reply src/osx/core/uilocale.mm:73]: > I did use `wxCFRef` originally, but I didn't want to have all the casts, so I got rid of it. But I've somehow completely missed `wxNSObjRef`, thanks. However it also takes ownership of the pointer in its ctor, so it's not really going to help here, is it?

right, also when using `wxNSObjRef`, we will have to `retain` explicitly, because these are not `create` Methods, however we could get rid of the `release` in our class.

> And I'm not sure I understand properties well enough to be sure I use them correctly. Should I just have `@property (retain)` here?

with pure c++ classes I don't think you can use properties.

> BTW, I build with ASAN (`-fsanitize=address`) under Mac and yet it doesn't detect memory leaks somehow, I wonder what am I doing wrong -- or could it really just not work correctly under ARM?

I don't really know. You'd need a autorelease pool in place, when running these methods to detect it at all, so that destroying this memory pool would lead to a double delete detection.

[the remark this answers is not in the document]
Comment (vadz) [reply src/osx/core/uilocale.mm]: OK, for now I'll stay with the manual retain/release then.

**Comment by vadz** [comment]
I've addressed the comments above and rebased+forcepushed once again.

**Review by csomor — APPROVED** [review]
thanks a lot

**Comment by vadz** [comment]
BTW, leak sanitizer indeed doesn't work under macOS, it's confirmed by running a simple program from https://stackoverflow.com/questions/53456304/mac-os-leaks-sanitizer or even just setting `ASAN_OPTIONS=detect_leaks=1`. This is really disappointing, I guess I need to install upstream clang and use it instead of Apple one. Or will it not work then?

File: interface/wx/uilocale.h [hunk 10]
```diff
@@ -218,15 +218,36 @@ class wxLocaleIdent
...
+        Note that this value is only used under Unix systems and simply ignored
+        under the other ones.
+
+        Return reference to @this for method chaining.
+
+        @param charset
+            Charset is a string such as "UTF-8", "ISO855915" or "KOI8R".
+            Supported charsets depend on the implementation and installation.
+    */
+    wxLocaleIdent& Charset(const wxString& charset);
```
Comment (vslavik) [inline interface/wx/uilocale.h:280]: I don't think this should exist here, especially if it's for use with wxUILocale — any support *UI* toolkit operates on Unicode and (on Unix) UTF-8. That locale on some Unices includes charset is a legacy thing that makes little sense in a new API. Can't the utf8 think mentioned in the comment be tried transparently when setting the locale?

File: include/wx/localedefs.h [hunk 11]
```diff
@@ -108,6 +108,83 @@ struct WXDLLIMPEXP_BASE wxLanguageInfo
...
 
+// ----------------------------------------------------------------------------
+// wxLocaleIdent: allows to fully identify a locale under all platforms
+// ----------------------------------------------------------------------------
+
+class WXDLLIMPEXP_BASE wxLocaleIdent
+{
+public:
+    // Leave language empty
+    wxLocaleIdent() { }
```
Comment (vslavik) [inline include/wx/localedefs.h:119]: What does "empty language" mean? Some of the code here suggests it means the default, but the docs don't spell it out.

File: include/wx/localedefs.h [hunk 12]
```diff
@@ -108,6 +108,83 @@ struct WXDLLIMPEXP_BASE wxLanguageInfo
...
+
+    // Accessors for the individual fields.
+    const wxString& GetLanguage() const { return m_language; }
+    const wxString& GetRegion() const { return m_region; }
+    const wxString& GetScript() const { return m_script; }
+    const wxString& GetCharset() const { return m_charset; }
+    const wxString& GetModifier() const { return m_modifier; }
+
+    // Construct platform dependent name
+    wxString GetName() const;
```
Comment (vslavik) [inline include/wx/localedefs.h:172]: This can return an empty string; again the meaning of that is unclear.

File: include/wx/localedefs.h [hunk 13]
```diff
@@ -108,6 +108,83 @@ struct WXDLLIMPEXP_BASE wxLanguageInfo
...
+class WXDLLIMPEXP_BASE wxLocaleIdent
+{
+public:
+    // Leave language empty
+    wxLocaleIdent() { }
+
+    // Construct from language, i.e. a two-letter ISO 639-1 code (or a
+    // three-letter ISO 639-2 code if there is no ISO 639-1 code for this
+    // language).
+    wxLocaleIdent(const char* language)
```
Comment (vslavik) [inline include/wx/localedefs.h:124]: I think — and tried to argue for that in the previous wxUILocale discussion — it is a mistake to construct language/locale identifier in this way in a modern API. Let's use the de-facto standard these days, which is [BCP 47 language tags](https://en.wikipedia.org/wiki/IETF_language_tag) ([brief intro](https://www.w3.org/International/articles/language-tags/)).

By all means, have a class useful for *parsing* or converting language identifier, but the primary input should be the tag: `wxLocaleIdent("es-005")`, not `wxLocaleIdent("es").Region("005")`. As the API is now, it doesn't allow construction of the identifier from a portable form (e.g. BCP-47 in a string, obtained from some API to determine user preference) and would require parsing in user code for that. And it already doesn't support full BCP-47 (dialects, private subtags), even without accounting for future expansions of the tags.

File: include/wx/private/uilocale.h [hunk 14]
```diff
@@ -12,6 +12,15 @@
 
 #include "wx/localedefs.h"
 #include "wx/string.h"
+#include "wx/vector.h"
+
+typedef wxVector<wxLanguageInfo> wxLanguageInfos;
+
+// Return the vector of all known languages.
+const wxLanguageInfos& wxGetLanguageInfos();
```
Comment (vslavik) [inline include/wx/private/uilocale.h:20]: (Beware that the included database is incomplete and doesn't even cover all languages supported by e.g. modern Windows, so "all" is a bit of a stretch.)

File: include/wx/private/uilocale.h [hunk 15]
```diff
@@ -33,18 +42,37 @@ class wxUILocaleImpl
...
 
-    // This function exists only for wxLocale compatibility and sets the locale
-    // corresponding to the given language.
+    // Create locale object for the given locale.
+    //
+    // It may return NULL in case of failure.
+    static wxUILocaleImpl* CreateForLocale(const wxLocaleIdent& locId);
+
+    // This function exists only for wxLocale compatibility and creates the
+    // locale corresponding to the given language.
```
Comment (vslavik) [inline include/wx/private/uilocale.h:51]: Perhaps use stronger language and say that it shouldn't be used in new and/or user code?

**Review by vslavik — COMMENTED** [review]
I didn't review the MSW (or too closely, OSX either) implementation and focused on the API. It looks globally fine to me (and so does the base wxUILocale PR, which I read through now too).

The one thing I think is a mistake is the wxLocaleIdent API, and feel somewhat strongly that for its implicit constructor, language tag is the only sensible argument, not just the language subtag as it is implemented now.

File: interface/wx/uilocale.h (hunk 10 above)
Comment (vadz) [reply interface/wx/uilocale.h:280]: I was going to not have it initially, but then decided that we were going to need it to support parsing Unix locale strings and added it. And I think encodings such as KOI8 or Shift-JIS might still be used in some places, although I admit I am not really sure.

And we do already try UTF-8 (just as we did with `wxLocale`) under Unix, but we actually use this method in the implementation, so removing it would need changing that code and making it less nice (right now all functions take just `wxLocaleIdent`), so I'd rather keep it, unless you really, really hate it. But I'll update the comment to make it more dissuasive.

File: include/wx/localedefs.h (hunk 11 above)
Comment (vadz) [reply include/wx/localedefs.h:119]: Yes, it is "the default", I'll have to explain it better.

This is another place where I hesitated for a long time: on one hand, a "default" identifier is not really useful, especially because it's not clear whether it's a system default, user default or something else. But OTOH this is a value-like class, so it would be strange to not have a default ctor in it...

File: include/wx/localedefs.h (hunk 12 above)
Comment (vadz) [reply include/wx/localedefs.h:172]: I was thinking about renaming this to something like `GetPlatformName()` and will probably do it.

Also, this method is mostly useful for the implementation, i.e. used in the platform-specific code, but I thought we could expose it as long as we had it anyhow. Do you think we shouldn't?

File: include/wx/localedefs.h (hunk 13 above)
Comment (vadz) [reply include/wx/localedefs.h:124]: This is a (strict) subset of BCP 47 language tags. Using full tags is problematic because how are we supposed to use them under Unix? If we parse them ourselves, we'd still be limited to the same subset, wouldn't we? And even macOS uses `language_REGION` format according to Apple docs and not BCP 47 `language-REGION` (although maybe it accepts both?).

I don't like passing the full tag to the ctor because then the current API doesn't make sense at all, e.g. `wxLocaleIdent("es-005").Region("ES")` is inconsistent. We could add `wxLocaleIdent::FromBCP47(string)` static ctor, but this is longer and I don't know how useful is it going to be. The current API was the best readability/safety/functionality compromise I could find.

FWIW I also thought about (additionally) providing a ctor from `wxLanguage` because even though our language database is incomplete, at least you can't make typos in language constants, unlike in strings.

To summarize, if we provide a ctor from BCP 47 string, we need to parse it ourselves, at least under Unix, and we couldn't provide safer/more readable components access as right now. Do we really want to do this?

File: include/wx/private/uilocale.h (hunk 14 above)
Comment (vadz) [reply include/wx/private/uilocale.h:20]: It was supposed to be "known to wx", I'll update it.

File: include/wx/private/uilocale.h (hunk 15 above)
Comment (vadz) [reply include/wx/private/uilocale.h:51]: It's a private header, so it could only be used inside wx itself, but I'll update the comment.

File: include/wx/localedefs.h (hunk 13 above)
Comment (csomor) [reply include/wx/localedefs.h:124]: > And even macOS uses `language_REGION` format according to Apple docs and not BCP 47 `language-REGION` (although maybe it accepts both?).

macOS differentiates between Language and Locale IDs

https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPInternational/LanguageandLocaleIDs/LanguageandLocaleIDs.html#//apple_ref/doc/uid/10000171i-CH15-SW1

Comment (vadz) [reply include/wx/localedefs.h:124]: Yes, I should have mentioned that I read this and this is where I saw

> [language designator]-[script designator]_[region designator]

form for the locale identifiers. And we do need a locale, rather than language, identifier here.

**Comment by vadz** [comment]
To put it another way: by requiring constructing the locale ID from the different parts, we limit the locales that we support, but we can ensure that we do support them on all platforms. By allowing construction from an arbitrary string, we wouldn't limit anything, but neither could we guarantee any degree of support.

E.g. right now you can construct a locale corresponding to `zh-Hant` and it will work on all 3 platforms, even though it will ignore the script part under Linux. Of course, we could also make it work by parsing the string ourselves, but this requires knowing about this part of the spec. OTOH the current code doesn't support extlangs (because I decided it wasn't worth it in practice, as they all have the corresponding "preferred value" anyhow) and so we don't support `ar-afb` -- but neither could we support it (under Linux and nor macOS, as I've just checked that this is not a supported locale under Big Sur) without adding explicit support for parsing extlangs even if we used string-based API.

If anything, I suggest adding a way to explicitly set an opaque *platform-dependent* string in `wxLocaleIdent`, so that you could request a "zh_HK BIG5-HKSCS" locale under Unix and "zh_Hans_HK" under macOS if you really wanted to. But I still believe we should be constructing such string from its constituent parts in the simple/default case because I want to just write `wxUILocale("fr")`.

@vslavik do you find this convincing or not at all?

File: include/wx/localedefs.h (hunk 11 above)
Comment (vslavik) [reply include/wx/localedefs.h:119]: > This is another place where I hesitated for a long time: on one hand, a "default" identifier is not really useful, especially because it's not clear whether it's a system default, user default or something else.

With (hopefully) better understanding now of what you're doing with this class, I'd say the default value semantics should definitely be "user default" and that a distinct default value is *necessary*. If a) wxLocaleIdent is used internally and b) it cannot, by design, represent all languages/locales, then a distinct way to represent user default is necessary.

File: include/wx/localedefs.h (hunk 12 above)
Comment (vslavik) [reply include/wx/localedefs.h:172]: No, having a way to get platform name is certainly useful.

File: include/wx/localedefs.h (hunk 13 above)
Comment (vslavik) [reply include/wx/localedefs.h:124]: > Using full tags is problematic because how are we supposed to use them under Unix?

Parse the information you can use from them, ignore (but preserve) the rest.

> I don't like passing the full tag to the ctor because then the current API doesn't make sense at all,

That much I can agree with, but my conclusion from that is different: get rid of the .Region() etc. stuff. *That* makes little sense to me. Locales are not a thing you routinely hardcode like this.

> e.g. wxLocaleIdent("es-005").Region("ES") is inconsistent.

Is `wxLocaleIdent("es").Region("005").Region("ES")` consistent?

> We could add wxLocaleIdent::FromBCP47(string) static ctor, but this is longer and I don't know how useful is it going to be. 

I'd say more than wxLocaleIdent(language). I think this is part of my own confusion w.r.t. language vs locale in this API that I demonstrated. Language only is rarely a sufficient *locale* identification for UI-visible aspects (e.g. date formatting differs between fr-FR and fr-CA).

> FWIW I also thought about (additionally) providing a ctor from wxLanguage because even though our language database is incomplete, at least you can't make typos in language constants, unlike in strings.

I have no problem with that, or the argument for having it, but would like to point out the inconsistency between  `wxLocaleIdent(wxLANGUAGE_FRENCH_CANADIAN)` being fine, but `wxLocaleIdent("fr-CA")` not.

> To summarize, if we provide a ctor from BCP 47 string, we need to parse it ourselves, at least under Unix, and we couldn't provide safer/more readable components access as right now. Do we really want to do this?

We could provide *read* components access (by parsing). I don't think *write* access makes sense, but I'm probably missing something. 

More importantly, I think it is natural to read `wxLocaleIdent("foo")` (or just `"foo"` because the ctor is implicit) as meaning `"foo"` is the locale. But it isn't in this API, it is only the language, and `wxLocaleIdent("de-AT")` fails. This is unintuitive to me and I think it will lead to errors. Note that from this point of view, `wxLocaleIdent().Language("fr")` is fine. 

At the very least, the ctor should assert on presence of `-` or `_` in the language.

**Comment by vslavik** [comment]
> By allowing construction from an arbitrary string, we wouldn't limit anything, but neither could we guarantee any degree of support.

That's not quite true. We could use the recognized subtags, and ignore (but preserve and roundtrip) the rest, possibly improving support in the future w/o requiring any user code changes. In your `ar-afb` example, it would be reduced to `ar` when unsupported.

But I'm not going to argue this is wrong approach; I have a feeling I am just missing the intended purpose.

What *is* extremely important, though, that no information is lost about the *user default* locale, which cannot be fully represented, even now, with this `wxLocaleIdent` design. So there *must* be a clear "use user default" value that is never ever translated into explicit form.

**Comment by vadz** [comment]
I'll have to think more about this, but, just to explain, the presence of `Region()` etc is due to my general dislike of string-typed APIs, I always prefer to rely on the compiler as much as possible.

But it's true that not being able to use `wxLocaleIdent("fr-FR")` is error-prone. OTOH if we support this, we need to support `wxLocaleIdent("fr_FR")` too because I guarantee that people will use it, whatever BCP 47 says.

Also, I really like having `Charset()` (which is, again, used internally) and `Script()` methods instead of having to parse the string to extract them (if only in order to discard them, as I'd need to do with the script under Unix). So it seems like a compromise solution would be to get rid of `Region()` but keep the other methods, but wouldn't it be rather strange to handle these components of the BCP 47 string specially, but not the language and the region?

FWIW my use case is a program which analyses texts in English, German, French and quite a few more languages, including ones with quite different sorting order, such as Swedish, Czech, Turkish etc, and needs to sort lists of words in them correctly. And, at least so far, I don't really care about the region for any of these languages, which is why I thought it would be nice to separate language from region.

> What is extremely important, though, that no information is lost about the user default locale, which cannot be fully represented, even now, with this wxLocaleIdent design. So there must be a clear "use user default" value that is never ever translated into explicit form.

Yes, we do agree about this and `wxLocaleIdent()` does represent it (and so does `wxUILocale::GetCurrent()` after calling `UseDefault()` which I expect all applications using `wxUILocale` to do).

**Comment by vslavik** [comment]
> I'll have to think more about this, but, just to explain, the presence of Region() etc is due to my general dislike of string-typed APIs, I always prefer to rely on the compiler as much as possible.

I know, and it makes sense for things like `wxSizerFlags` that are hardcoded, but is locale of the same category? It apparently is in the program you're working on, but I see it more like a *data-driven* thing. It either comes from user configuration or from some data, rarely hard-coded. For example, I couldn't use this as-is for sorting columns in Poedit, because the language tag comes from the file and can be anything. So I'd need to parse it myself and *my* view is that this is more common situation than yours where you hardcode languages. I may be wrong.  But even e.g. supported languages for OCR or machine translation would typically come from some enumeration API, wouldn't they? 

> OTOH if we support this, we need to support wxLocaleIdent("fr_FR") too because I guarantee that people will use it, whatever BCP 47 says.

Right.

> So it seems like a compromise solution would be to get rid of Region() but keep the other methods, but wouldn't it be rather strange to handle these components of the BCP 47 string specially, but not the language and the region?

It would, such "compromise" wouldn't be good.

File: include/wx/localedefs.h (hunk 11 above)
Comment (vadz) [reply include/wx/localedefs.h:119]: OTOH we already have `wxUILocale::CreateUserDefault()` and having `wxUILocale(wxLocaleIdent())` do the same thing doesn't look like such a great idea.

I think I'd rather make default-constructed `wxLocaleIdent` invalid, i.e. assert that it's not empty in `wxUILocale` ctor. What do you think?

**Comment by vadz** [comment]
I think your argument is that we need a way to parse BCP 47 strings in wx and I agree with it. I wanted to postpone doing this until later because it's not something I need myself right now and I thought it could be just an addition, i.e. could be done without breaking backwards compatibility, but I'll do it now (in some form), as it doesn't seem that difficult and, at least, it will address this particular point.

The problem is that I still like constructing the locale identifier from parts too, but I can't see any reasonable API that allows both without being too confusing and error-prone. The best I can come up with right now is:

1. Remove `wxLocaleIdent()` non-default ctor, i.e. force people writing `wxLocaleIdent().Language("fr").Region("FR")` (this is also compatible with making empty `wxLocaleIdent` invalid, which is I think the right thing to do).
2. Add `wxUILocale` ctor from string, that will be interpreted as full BCP 47-like locale string. In fact, I'm probably going to add some static method instead, which is slightly more difficult as it will require making `wxUILocale` copyable (it isn't currently), but should be more clear, especially if someone can come up with a good name for this function.

Like this, in your situation you won't be using `wxLocaleIdent` at all and just operate with strings, while the programs that do hardcode the list of locales they support (and this definitely has to be the case for mine, because it needs to know plenty of other stuff about the languages it supports, so it can never support some arbitrary language it doesn't know about) could use `wxLocaleIdent` in a more type-safe way.

What do you think?

**Comment by vadz** [comment]
I'd really like to advance with this because I need it, so I've done the changes proposed above and pushed.

Please let me know if you think this is acceptable.

BTW, I tried to find some language tag allowed by BCP 47 but not expressible in `language-script-region` form supported by MSW, but couldn't. E.g. it claims to support `de-CH-1996`, but then it also claims to support `de-CH-1995` and `de-CH-1997`, so this doesn't seem very reliable...

**Comment by vslavik** [comment]
> The problem is that I still like constructing the locale identifier from parts too, but I can't see any reasonable API that allows both without being too confusing and error-prone. The best I can come up with right now is:

I still don't understand why you think having both is error-prone… `wxLocaleIdent("fr-FR").Region("CA")` is IMHO no more error-prone, or more confusing, than `wxLocaleIdent("fr").Region("FR").Region("CA")`. Both allow over-specifying subtags and in both the right-most value takes precedence.

But I agree having exactly one way to specify the region is is better than 2+; I think that's your point?

> Remove wxLocaleIdent() non-default ctor, i.e. force people writing wxLocaleIdent().Language("fr").Region("FR") (this is also compatible with making empty wxLocaleIdent invalid, which is I think the right thing to do).

Invalid still means "user default"?

> Add wxUILocale ctor from string, that will be interpreted as full BCP 47-like locale string. In fact, I'm probably going to add some static method instead, which is slightly more difficult as it will require making wxUILocale copyable (it isn't currently), but should be more clear, especially if someone can come up with a good name for this function.

I use `Language::FromLanguageTag()` in Poedit, ICU uses `icu_forLanguageTag()`. 

> What do you think?

Sounds good to me, thanks!

**Comment by vadz** [comment]
> > The problem is that I still like constructing the locale identifier from parts too, but I can't see any reasonable API that allows both without being too confusing and error-prone. The best I can come up with right now is:
> 
> I still don't understand why you think having both is error-prone… `wxLocaleIdent("fr-FR").Region("CA")` is IMHO no more error-prone, or more confusing, than `wxLocaleIdent("fr").Region("FR").Region("CA")`. Both allow over-specifying subtags and in both the right-most value takes precedence.

I agree that the two shown above are not very different, but I think that using the default ctor, as the latest version does, is more clear and less error-prone as you're much less likely to call the same function twice.
 
> But I agree having exactly one way to specify the region is is better than 2+; I think that's your point?

Yes, which is why I removed the possibility to specify it in the ctor and added a separate function for this.

> > Remove wxLocaleIdent() non-default ctor, i.e. force people writing wxLocaleIdent().Language("fr").Region("FR") (this is also compatible with making empty wxLocaleIdent invalid, which is I think the right thing to do).
> 
> Invalid still means "user default"?

No, invalid means empty and can't be used for anything. The default user locale is accessed via `wxUILocale::GetDefault()` as I think it's preferable to be explicit about it. If we really need to represent it at `wxLocaleIdent` level, we could add a separate `Default()` static ctor for it too, but right now this doesn't seem to be necessary.
 
> > Add wxUILocale ctor from string, that will be interpreted as full BCP 47-like locale string. In fact, I'm probably going to add some static method instead, which is slightly more difficult as it will require making wxUILocale copyable (it isn't currently), but should be more clear, especially if someone can come up with a good name for this function.
> 
> I use `Language::FromLanguageTag()` in Poedit, ICU uses `icu_forLanguageTag()`.

Thanks, I didn't even think to check the ICU name (let alone Poedit one), but I used `FromTag()`, which seems like a good enough name in this context.

> > What do you think?
> 
> Sounds good to me, thanks!

Thanks for the reviews and advice! I'm going to merge it, but more changes remain, of course, possible, please don't hesitate to propose them if you see that something is missing for its use in Poedit. TIA!
